### PR TITLE
xc7 MMCM support

### PR DESCRIPTION
**Pull request by mkurc-ant**
This pull requests adds support for `MMCM` block for Xilinx 7-series devices.

The supported features are:
- Basic (integer) multiplier and divider control,
- Fractional multiplier (`CLKFBOUT`) and divider (`CLKOUT0`) support,
- Duty cycle control,
- Phase shift control,
- Bandwidth and compensation settings,
- Local control input inverters (`CLKINSEL`, `PSEN`, `PSINCDEC`, `PWRDWN`, `RST`)
- Startup wait enable/disable

Dynamic phase shift and dynamic reconfiguration are supported as well.

All the remaining `MMCM` functionality is not supported at the moment mostly due to missing bitstream documentation. This is tracked in https://github.com/SymbiFlow/prjxray/issues/1473

This PR also adds tests/examples for `MMCM` which are available under `xc/xc7/tests/mmcm` and can be run on hardware (some of them).

### CODE REVIEW

**Comment by mkurc-ant** [comment]
While disassembling Vivado bitstreams of MMCM designs I'm consistently getting these two unknown bits:
```
{ unknown_bit = "0040151c_30_19", unknown_segment = "0x00401500", unknown_segbit = "28_979" }
{ unknown_bit = "0040151c_31_28", unknown_segment = "0x00401500", unknown_segbit = "28_1020" }
```

**Comment by litghost** [comment]
> While disassembling Vivado bitstreams of MMCM designs I'm consistently getting these two unknown bits:
> 
> ```
> { unknown_bit = "0040151c_30_19", unknown_segment = "0x00401500", unknown_segbit = "28_979" }
> { unknown_bit = "0040151c_31_28", unknown_segment = "0x00401500", unknown_segbit = "28_1020" }
> ```

This doesn't surprise me too much.  I disabled the compension features because they were initially causing me trouble, see here: https://github.com/SymbiFlow/prjxray/blob/91d91357b5077b12e578ac741322bea71c285b6a/fuzzers/031-cmt-mmcm/generate.py#L128-L131

Now that the basic MMCM features and pips are in, maybe it is time to go back and get the compension features working?

**Comment by mkurc-ant** [comment]
@litghost I've checked and currently the `MMCM` does not work in hardware when using SymbiFlow bitstream. It looks like the two unknown bits I mentioned before are required for it to work. I've tested Vivado and SymbiFlow bitstreams for the same design where the only difference between them was the latter missing these two bits. The Vivado bitstream works while the SymbiFlow doesn't.

**Comment by mkurc-ant** [comment]
This one solves the two missing bits: https://github.com/SymbiFlow/prjxray/pull/1491

File: xc/xc7/archs/artix7/tiles/mmcme2_adv/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+project_ray_tile(
+  ARCH artix7
+  TILE MMCME2_ADV
+  SITE_AS_TILE
+  SITE_TYPES MMCME2_ADV/MMCME2_ADV
+  EQUIVALENT_SITES SLICEL
```
Comment (litghost) [inline xc/xc7/archs/artix7/tiles/mmcme2_adv/CMakeLists.txt:6]: This equivalent site is wrong?

File: xc/xc7/tests/mmcm/gen_random_cases.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+
+        if f_vco < vco_range[0]:
+            return False
+        if f_vco > vco_range[1]:
+            return False
+
+    return True
+
+
+def make_integers(params):
```
Comment (litghost) [inline xc/xc7/tests/mmcm/gen_random_cases.py:157]: This function shouldn't be needed here any more?

**Comment by litghost** [comment]
@mkurc-ant Can you please rebase this PR?

**Comment by mkurc-ant** [comment]
@litghost Done. I'm now working on adding MMCM support to fasm2bels.

File: xc/xc7/tests/mmcm/gen_random_cases.py (hunk 2 above)
Comment (mkurc-ant) [reply xc/xc7/tests/mmcm/gen_random_cases.py]: I've removed all code that converts MMCM parameters.

**Comment by litghost** [comment]
Latest error:

```
CMake Error at xc/xc7/tests/mmcm/CMakeLists.txt:46 (add_dependencies):
  The dependency target "mmcm_random_case3_vivado_diff_fasm" of target
  "all_xc7_mmcm_vivado_diff_fasm" does not exist.
```

File: xc/xc7/archs/artix7/tiles/mmcme2_adv/CMakeLists.txt (hunk 1 above)
Comment (mkurc-ant) [reply xc/xc7/archs/artix7/tiles/mmcme2_adv/CMakeLists.txt]: Fixed.

**Review by litghost — CHANGES_REQUESTED** [review]
Looks like there are 3 outstanding issues.

PLL `NO_COUNT` parameter mismatch, which appears to be a new regression:
```
FAILED: cd /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7-vendor/build/xc/xc7/tests/idelayctrl && diff -u /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7-vendor/build/xc/xc7/tests/idelayctrl/idelayctrl_arty/artix7-xc7a50t-virt-xc7a50t-test/top.bit.fasm /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7-vendor/build/xc/xc7/tests/idelayctrl/idelayctrl_arty/artix7-xc7a50t-virt-xc7a50t-test/interchange/idelayctrl_arty_vivado_interchange.bit.fasm
--- /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7-vendor/build/xc/xc7/tests/idelayctrl/idelayctrl_arty/artix7-xc7a50t-virt-xc7a50t-test/top.bit.fasm	2021-01-21 08:46:31.989570289 -0800
+++ /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7-vendor/build/xc/xc7/tests/idelayctrl/idelayctrl_arty/artix7-xc7a50t-virt-xc7a50t-test/interchange/idelayctrl_arty_vivado_interchange.bit.fasm	2021-01-21 08:47:44.570461160 -0800
@@ -213,14 +213,12 @@
 CMT_TOP_L_UPPER_T_X106Y44.PLLE2.CLKOUT0_CLKOUT1_HIGH_TIME[0]
 CMT_TOP_L_UPPER_T_X106Y44.PLLE2.CLKOUT0_CLKOUT1_LOW_TIME[0]
 CMT_TOP_L_UPPER_T_X106Y44.PLLE2.CLKOUT0_CLKOUT1_OUTPUT_ENABLE[0]
-CMT_TOP_L_UPPER_T_X106Y44.PLLE2.CLKOUT0_CLKOUT2_NO_COUNT[0]
 CMT_TOP_L_UPPER_T_X106Y44.PLLE2.CLKOUT1_CLKOUT1_HIGH_TIME[1]
 CMT_TOP_L_UPPER_T_X106Y44.PLLE2.CLKOUT1_CLKOUT1_LOW_TIME[1]
 CMT_TOP_L_UPPER_T_X106Y44.PLLE2.CLKOUT1_CLKOUT1_OUTPUT_ENABLE[0]
 CMT_TOP_L_UPPER_T_X106Y44.PLLE2.CLKOUT2_CLKOUT1_HIGH_TIME[0]
 CMT_TOP_L_UPPER_T_X106Y44.PLLE2.CLKOUT2_CLKOUT1_LOW_TIME[0]
 CMT_TOP_L_UPPER_T_X106Y44.PLLE2.CLKOUT2_CLKOUT1_OUTPUT_ENABLE[0]
-CMT_TOP_L_UPPER_T_X106Y44.PLLE2.CLKOUT2_CLKOUT2_NO_COUNT[0]
 CMT_TOP_L_UPPER_T_X106Y44.PLLE2.CLKOUT3_CLKOUT1_HIGH_TIME[0]
 CMT_TOP_L_UPPER_T_X106Y44.PLLE2.CLKOUT3_CLKOUT1_LOW_TIME[0]
 CMT_TOP_L_UPPER_T_X106Y44.PLLE2.CLKOUT3_CLKOUT2_NO_COUNT[0]
```

MMCM DCP parameter failures:
```
FAILED: cd /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7-vendor/build/xc/xc7/tests/mmcm/mmcm_int_frac_basys3/artix7-xc7a50t-bottom-virt-xc7a50t-bottom-test/interchange && /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7-vendor/env/conda/envs/symbiflow_arch_def_base/bin/cmake -E env XRAY_VIVADO_SETTINGS=/image/Xilinx/Vivado/2019.2/settings64.sh /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7-vendor/third_party/prjxray/utils/vivado.sh -mode batch -source /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7-vendor/build/xc/xc7/tests/mmcm/mmcm_int_frac_basys3/artix7-xc7a50t-bottom-virt-xc7a50t-bottom-test/interchange/mmcm_int_frac_basys3_vivado_interchange_runme.tcl > /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7-vendor/build/xc/xc7/tests/mmcm/mmcm_int_frac_basys3/artix7-xc7a50t-bottom-virt-xc7a50t-bottom-test/interchange/vivado.stdout.log
ERROR: [DRC ADEF-499] SS_MOD_PERIOD_defined: Cell mmcme2_test.mmcmCMT_TOP_L_LOWER_B_X106Y9_MMCME2_ADV_X1Y0_MMCME2_ADV has no value for attribute SS_MOD_PERIOD. For proper hardware operation, SS_MOD_PERIOD must be assigned a decimal value. To correct this issue, modify the design for this cell.
ERROR: [DRC AVAL-38] V7V8V9_MMCM_SET_COMPENSATION3: Cell mmcme2_test.mmcmCMT_TOP_L_LOWER_B_X106Y9_MMCME2_ADV_X1Y0_MMCME2_ADV of type 'MMCME2_ADV' has been detected to have a configuration that requires the COMPENSATION setting to be INTERNAL. The application can automatically set the COMPENSATION attribute provided that the original COMPENSATION attribute is left with its default value by the designer.
ERROR: [DRC PLCR-1] Placement Constraints Check for Clock Region(s): Design Check found an error in Clock Region X0Y0. This clock region has 14 clocks locked whereas only 12 clocks can be routed per clock region. The cells were either constrained through Area Groups or locked via user constraints. This situation can be resolved by unlocking some cells from the user constraints so that there are only 12 clocks in this region.
ERROR: [Vivado 12-1345] Error(s) found during DRC. Bitgen not run.
```

BUFH needs to be a passthrough, rather than an explicit cell:
```
FAILED: cd /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7-vendor/build/xc/xc7/tests/mmcm/mmcm_int_frac_basys3/artix7-xc7a50t-bottom-virt-xc7a50t-bottom-test && /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7-vendor/env/conda/envs/symbiflow_arch_def_base/bin/cmake -E remove -f mmcm_int_frac_basys3/artix7-xc7a50t-bottom-virt-xc7a50t-bottom-test/design_mmcm_int_frac_basys3_vivado.dcp && /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7-vendor/env/conda/envs/symbiflow_arch_def_base/bin/cmake -E remove -f mmcm_int_frac_basys3/artix7-xc7a50t-bottom-virt-xc7a50t-bottom-test/design_mmcm_int_frac_basys3_vivado.xpr && /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7-vendor/third_party/prjxray/utils/vivado.sh -mode batch -source /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7-vendor/build/xc/xc7/tests/mmcm/mmcm_int_frac_basys3_vivado_runme.tcl > /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7-vendor/build/xc/xc7/tests/mmcm/mmcm_int_frac_basys3/artix7-xc7a50t-bottom-virt-xc7a50t-bottom-test/vivado.stdout.log
ERROR: [DRC PLCR-1] Placement Constraints Check for Clock Region(s): Design Check found an error in Clock Region X0Y0. This clock region has 14 clocks locked whereas only 12 clocks can be routed per clock region. The cells were either constrained through Area Groups or locked via user constraints. This situation can be resolved by unlocking some cells from the user constraints so that there are only 12 clocks in this region.
ERROR: [Vivado 12-1345] Error(s) found during DRC. Bitgen not run.
[3291/3547] Generating mmcm_int_basys3/artix7-xc7a50t-bottom-virt-xc7a50t-bottom-test/interchange/mmcm_int_basys3_vivado_interchange.dcp
```

**Comment by mkurc-ant** [comment]
@litghost Thanks for pointing out all the issues. So far I've managed to fix PLL `NO_COUNT` regression and MMCM DCP errors. A PR is in: https://github.com/SymbiFlow/symbiflow-xc-fasm2bels/pull/44

**Comment by mithro** [comment]
Any reason this pull request was never landed?

**Comment by mkurc-ant** [comment]
@mithro As far as I remember this didn't land because there are some MMCM tests that fail due to incorrect BUFHCE route-throu handling in fasm2bels. I see two options:
 - We solve the BUFHCE issue in fasm2bels,
 - We simplify the failing tests so that they use eg. a single clock output which does not cause issues.

**Comment by mkurc-ant** [comment]
I've tested locally integration of this PR with https://github.com/SymbiFlow/symbiflow-xc-fasm2bels/pull/52 and the latter still fails on some designs each time at the same place in code:
```
Traceback (most recent call last):
  File "/home/build/mkurc/symbiflow-arch-defs/env/conda/envs/symbiflow_arch_def_base/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/build/mkurc/symbiflow-arch-defs/env/conda/envs/symbiflow_arch_def_base/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/build/mkurc/symbiflow-arch-defs/third_party/symbiflow-xc-fasm2bels/fasm2bels/__main__.py", line 15, in <module>
    main()
  File "/home/build/mkurc/symbiflow-arch-defs/third_party/symbiflow-xc-fasm2bels/fasm2bels/fasm2bels.py", line 489, in main
    top.make_routes(allow_orphan_sinks=args.allow_orphan_sinks)
  File "/home/build/mkurc/symbiflow-arch-defs/third_party/symbiflow-xc-fasm2bels/fasm2bels/models/verilog_modeling.py", line 2174, in make_routes
    self.handle_post_route_cleanup()
  File "/home/build/mkurc/symbiflow-arch-defs/third_party/symbiflow-xc-fasm2bels/fasm2bels/models/verilog_modeling.py", line 2347, in handle_post_route_cleanup
    site.post_route_cleanup(self, site)
  File "/home/build/mkurc/symbiflow-arch-defs/third_party/symbiflow-xc-fasm2bels/fasm2bels/models/clk_models.py", line 191, in cleanup_hrow
    assert False, bufhce_i_wire_pkey
AssertionError: 1627284
```
So it looks like BUFHCE route-throu handling still fails for some cases.

**Comment by mkurc-ant** [comment]
xc7 tests fail on `minilitex_ddr_arty` with:
```
ERROR: [DRC RTRES-2] Global clock nets are using local routing resources: Global clock nets are using local routing resources. 1 net(s) have at least one node with COST_CODE_NAME equal to BOUNCEACROSS or INTENT_CODE_NAME equal to NODE_DOUBLE, NODE_HLONG, NODE_HQUAD, NODE_SINGLE, NODE_VLONG, or NODE_VQUAD. This situation occurs when a global clock net must enter a clock region where all global clock routing resources are occupied. Check the report from report_clock_utilization to determine which clock regions are traversed by the problem net and use floorplanning or other physical constraints to ensure a maximum of 12 global clock nets occupy each clock region. The problem net(s) are sys_clk.
ERROR: [Vivado 12-1345] Error(s) found during DRC. Bitgen not run.
```

**Comment by acomodi** [comment]
I believe this error is not related to the change in fasm2bels, but it is a consequence of not handling the clock nets limits in a clock region, and this has not to do with fasm2bels, but with VPR.

If the MMCM test passed in the vendor tools CI I think we can go ahead and merge https://github.com/SymbiFlow/symbiflow-xc-fasm2bels/pull/52 and than update the fasm2bels version in this PR

**Comment by mkurc-ant** [comment]
@acomodi Ok, go ahead then with merging the fasm2bels PR.

I haven't looked in Vivado documentation what this DRC error means but it looks like that it is merely a warning that a clock uses a local route instead of a global one (because there are no more available).

**Pull request merged**